plugins/modules/cloud/misc/terraform.py

```diff
@@ -449,12 +480,72 @@ def main():
     if state == 'present' and module.params.get('parallelism') is not None:
         command.append('-parallelism=%d' % module.params.get('parallelism'))
 
+    def format_args(vars):
+        if isinstance(vars, str):
+            return '"{string}"'.format(string=vars)
```

Comment:
I think you should better use `json.dumps()` here, otherwise there will be problems with properly escaping quotes and other things.

Reply:
Although `json.dumps()` do encode special characters, so is the ```shlex_quote()```. Terraform did not like the double encapsulation, although the code/commands work without fail, it did not pass the correct string from ansible to terraform. I created a test case in  https://github.com/kosalaat/terrafrom-tests using terraform null_resource, which now test all the scenarios, including checking whether string are passed in intact.

Is there a way we can make this part of the repo?

Reply:
For the above TF workfiles, the following ansible playbooks works:

```
---

- hosts: localhost
  gather_facts: false
  tasks:
    - community.general.terraform:
        project_path: "../../terraform/null-resource"
        state: present
        variables:
          dictionaries:
            name: kosala
            age: 44
          list_of_strings:
            - kosala
            - neloufer
            - punu
            - ari
          list_of_objects:
            - name: kosala
              age: 44
            - name: neloufer
              age: 44
            - name: punu
              age: 9
            - name: ari
              age: 1
          boolean: true
          string_type: "randomstring2&$%@"
          list_of_lists:
            - [ 1 ]
            - [ 11, 12, 13 ]
            - [ 2 ]
            - [ 3 ]
```

Reply:
> Although `json.dumps()` do encode special characters, so is the `shlex_quote()`. Terraform did not like the double encapsulation,

I'm not sure what you mean. There is no `shlex_quote()` involved anywhere (with the variables you are touching).

Reply:
What if `string` happens to be `a"b` here? Then terraform will get passed `"a"b"`. I'm pretty sure it will **not** be happy with that.

Reply:
No it's not. `shlex_quote()` inside `run_command()` will sort that out.

Reply:
I'm still wondering how strings containing `"` will be handled. Simply adding quotes around them is probably not ok?

Reply:
You want to look at https://www.terraform.io/language/expressions/strings#escape-sequences to see how to correctly quote strings.